agot-bg-game-server/src/common/ingame-game-state/game-data-structure/Game.ts

```diff
@@ -80,6 +81,16 @@ export default class Game {
         return result;
     }
 
+    get nextWesterosCardTypes(): WesterosCardType[][] {
+        const result: WesterosCardType[][] = [];
+
+        this.westerosDecks.forEach(wd => {
+            result.push(wd.slice(0, this.revealedWesterosCards).map((card) => card.type));
```

Comment:
> 
> 
> I just realized that this will be problematic during end game. In round 9 it will still show 3 next cards, although there are just 2. The current one and the one of round 10. That's because we discard cards only by setting the discarded flag. So we have to apply a filter on `wd` for non discarded cards here.

**Edit** discarded won't be enough due to Winter is Coming. See later comments. We have to calculate `remainingRounds` and take them into account.

Reply:
I didn't understand the edit. A filter for non-discarded should work perfectly here. Winter is coming is not rsolved immediately. It is resolved on the turn it should be resolved, so that wouldn't affect the results. In fact, the rules state that the top 3 cards should stay revealed at any point during the game, including rounds 9 and 10 (only if it's not possible, like in deck 3, then there's nothing we can do). 

Making it so that on round 10 only the next card is revealed could be problematic if, for example, a WiC is drawn during the previous rounds and a 2nd WiC is drawn on round 10. If a WiC is drawn during round 10, we still know that the 2nd and 3rd card will not be shuffled (so it helps with the randomness of WiC, realising that some of the cards will not be shuffled). If we took into account the number of rounds, which CoK manual doesn't ask us to do, then a WiC on round 10 would be more random than a WiC on the rounds before it.

Reply:
I don't have much time right now.

But

> I didn't understand the edit. A filter for non-discarded should work perfectly here. Winter is coming is not rsolved immediately. It is resolved on the turn it should be resolved, so that wouldn't affect the results. 

Is not true. It's clarified by errata, that WiC is resolved immediately after it's drawn so you know what event is coming on deck 2 and that's how it's implemented. My concern is, when we only apply a filter for non-discarded cards, a WiC in round 9 will undiscard all cards again which will then again show 3 upcoming Westeros cards, though there is just 1! I want to avoid that as it might lead to confusion to the players.

Reply:
>If we took into account the number of rounds, which CoK manual doesn't ask us to do, then a WiC on round 10 would be more random than a WiC on the rounds before it.

Well in fact that is a good question for Jason (the game designer). I expect a WiC in round 10 to be more random than a WiC before. In round 10 even with this option I would shuffle all cards as from my point of view there aren't 2 additional upcoming cards revealed. Let's consider no WiC until round 10 and it comes in round 10... At that point there is only "upcoming card" left which would not be shuffled by your thought (10 cards in deck minus 9 Westeros rounds). Of course it's a legal thought and I guess we should discuss it in Discord how it should be handled.

Reply:
Ok according to the fact that rulebook says there is no discard pile I guess we can leave it as it is even if it is kind of weird to me. Sorry for the trouble and thx for the smart implementation.